File: android/src/main/kotlin/com/compress/all/flutter_compress_pro/ImageEngine.kt
package com.compress.all.flutter_compress

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import android.os.Build
import android.util.Log
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File

/**
 * Native image compression via the Android imaging stack (BitmapFactory +
 * Bitmap.compress). No FFmpeg, no Dart software decode. For a target size it
 * searches the quality (and downscales if needed) — images encode in
 * milliseconds, so this lands accurately at/under the target.
 *
 * Runs on a background dispatcher (see `FlutterCompressPlugin.dispatch`); every
 * intermediate bitmap is recycled so batches don't accumulate native heap.
 */
class ImageEngine(private val context: Context) {

    fun info(path: String): Map<String, Any?> {
        val src = probe(Src.OfFile(path))
        return mapOf(
            "path" to path,
            "width" to src.width,
            "height" to src.height,
            "sizeBytes" to File(path).length(),
            "format" to src.format,
        )
    }

    /** One encode's product, before it becomes a file or a byte array. */
    private class EncodeOutcome(
        val bytes: ByteArray,
        val width: Int,
        val height: Int,
        val format: String,
        val keptOriginal: Boolean,
        val src: Probe,
    )

    fun compress(
        path: String,
        config: ImageConfig,
        outputDir: String?,
        outputName: String?,
    ): Map<String, Any?> {
        val originalSize = File(path).length()
        val out = encode(Src.OfFile(path), originalSize, config)
        if (out.keptOriginal) {
            return mapOf(
                "outputPath" to path,
                "originalSizeBytes" to originalSize,
                "compressedSizeBytes" to originalSize,
                "width" to out.src.width,
                "height" to out.src.height,
                "format" to out.src.format,
                "skipped" to true,
            )
        }
        val file = context.resolveOutput(outputDir, outputName, path, ext(out.format))
        file.writeBytes(out.bytes)
        // EXIF is appended after the size search, so it can add a few hundred
        // bytes on top of targetSizeKB; the size reported below is the real one.
        if (config.keepExif && out.format == "jpeg") copyExif(path, file.absolutePath)
        return mapOf(
            "outputPath" to file.absolutePath,
            "originalSizeBytes" to originalSize,
            "compressedSizeBytes" to file.length(),
            "width" to out.width,
            "height" to out.height,
            "format" to out.format,
            "skipped" to false,
        )
    }

    /**
     * In-memory variant: nothing is read from or written to disk — except when
     * [ImageConfig.keepExif] is set, because the framework `ExifInterface` can
     * only `saveAttributes()` to a file or file descriptor. That opt-in path
     * borrows a temp file and deletes it again.
     */
    fun compressBytes(source: ByteArray, config: ImageConfig): Map<String, Any?> {
        val out = encode(Src.OfBytes(source), source.size.toLong(), config)
        if (out.keptOriginal) {
            return mapOf(
                "bytes" to source,
                "originalSizeBytes" to source.size.toLong(),
                "width" to out.src.width,
                "height" to out.src.height,
                "format" to out.src.format,
                "skipped" to true,
            )
        }
        var bytes = out.bytes
        if (config.keepExif && out.format == "jpeg") {
            bytes = withExifFromBytes(source, bytes)
        }
        return mapOf(
            "bytes" to bytes,
            "originalSizeBytes" to source.size.toLong(),
            "width" to out.width,
            "height" to out.height,
            "format" to out.format,
            "skipped" to false,
        )
    }

    /**
     * Round-trip through a temp file so `ExifInterface.saveAttributes()` has
     * something seekable to write into, then hand the bytes back.
     */
    private fun withExifFromBytes(source: ByteArray, encoded: ByteArray): ByteArray {
        val srcTmp = File(context.compressCacheDir(), "exif_src_${System.nanoTime()}.jpg")
        val dstTmp = File(context.compressCacheDir(), "exif_dst_${System.nanoTime()}.jpg")
        return try {
            srcTmp.writeBytes(source)
            dstTmp.writeBytes(encoded)
            copyExif(srcTmp.absolutePath, dstTmp.absolutePath)
            dstTmp.readBytes()
        } catch (e: Throwable) {
            // Metadata is best-effort; never fail the compression over it.
            Log.w(TAG, "keepExif: could not attach EXIF to in-memory output", e)
            encoded
        } finally {
            srcTmp.delete()
            dstTmp.delete()
        }
    }

    private fun encode(
        source: Src,
        originalSize: Long,
        config: ImageConfig,
    ): EncodeOutcome {
        val src = probe(source)

        // A null format keeps the source's format.
        val effective = resolveFormat(config.format ?: src.format)
        // PNG is inherently lossless; lossless mode also disables quality search.
        val lossy = !config.lossless && (effective == "jpeg" || effective == "webp")
        // A target size can't be honored losslessly — ignore it in that case.
        val targetBytes = if (config.lossless) null else config.targetSizeKB?.let { it * 1024 }

        var bmp = decodeUpright(source, src, config.maxWidth, config.maxHeight)
        try {
            val bytes: ByteArray
            if (targetBytes == null) {
                bytes = encode(bmp, effective, if (lossy) config.quality else MAX_QUALITY, config.lossless)
            } else {
                var fit = fitToTarget(bmp, effective, lossy, targetBytes, MIN_QUALITY)
                var tries = 0
                while (fit.bytes.size > targetBytes && tries < MAX_DOWNSCALES && bmp.width > MIN_SIDE) {
                    bmp = bmp.scaledTo(bmp.width * 3 / 4, bmp.height * 3 / 4)
                    // A smaller image fits at least the quality the last round
                    // reached, so the search can start from there.
                    fit = fitToTarget(bmp, effective, lossy, targetBytes, fit.quality)
                    tries++
                }
                bytes = fit.bytes
            }

            // Re-encoding can end up larger than the source (already-compressed
            // input, or lossless). If so, the caller gets the original back.
            val keptOriginal = SizeMath.keepsOriginal(
                bytes.size.toLong(), originalSize,
                config.keepOriginalIfLarger, config.minSavingsPercent,
            )
            return EncodeOutcome(
                bytes = bytes,
                width = bmp.width,
                height = bmp.height,
                format = effective,
                keptOriginal = keptOriginal,
                src = src,
            )
        } finally {
            bmp.recycle()
        }
    }

    // ---- source inspection -------------------------------------------------

    /**
     * Where the pixels come from. `BitmapFactory` and `ExifInterface` both have
     * separate file and byte-array entry points, so this is the only thing the
     * decode path needs to vary between `compress` and `compressBytes`.
     */
    private sealed interface Src {
        fun decode(opts: BitmapFactory.Options): Bitmap?
        fun orientation(): Int

        class OfFile(val path: String) : Src {
            override fun decode(opts: BitmapFactory.Options): Bitmap? =
                BitmapFactory.decodeFile(path, opts)

            override fun orientation(): Int = runCatching {
                ExifInterface(path)
                    .getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)
            }.getOrDefault(ExifInterface.ORIENTATION_NORMAL)
        }

        class OfBytes(val bytes: ByteArray) : Src {
            override fun decode(opts: BitmapFactory.Options): Bitmap? =
                BitmapFactory.decodeByteArray(bytes, 0, bytes.size, opts)

            override fun orientation(): Int = runCatching {
                ExifInterface(ByteArrayInputStream(bytes))
                    .getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)
            }.getOrDefault(ExifInterface.ORIENTATION_NORMAL)
        }
    }

    /** Source dimensions (already oriented), format and raw EXIF orientation. */
    private class Probe(
        val width: Int,
        val height: Int,
        val format: String,
        val orientation: Int,
    )

    /**
     * One bounds-only decode for everything we need up front. Dimensions are
     * reported **oriented** (a portrait photo stored as landscape + rotate-90
     * reads as portrait), matching what the user sees.
     */
    private fun probe(src: Src): Probe {
        val opts = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        src.decode(opts)
        if (opts.outWidth <= 0 || opts.outHeight <= 0) error("Could not read image")
        val orientation = src.orientation()
        val swap = orientation == ExifInterface.ORIENTATION_ROTATE_90 ||
            orientation == ExifInterface.ORIENTATION_ROTATE_270 ||
            orientation == ExifInterface.ORIENTATION_TRANSPOSE ||
            orientation == ExifInterface.ORIENTATION_TRANSVERSE
        val format = when (opts.outMimeType?.removePrefix("image/")?.lowercase()) {
            "png" -> "png"
            "webp" -> "webp"
            "heic", "heif" -> "heic"
            else -> "jpeg"
        }
        return Probe(
            width = if (swap) opts.outHeight else opts.outWidth,
            height = if (swap) opts.outWidth else opts.outHeight,
            format = format,
            orientation = orientation,
        )
    }

    /**
     * The format actually encoded. The request is always kept — JPEG stays JPEG,
     * PNG stays PNG, WebP stays WebP — except HEIC, which Android's Bitmap stack
     * can't encode, so it falls back to JPEG. (JPEG has no lossless mode; in
     * lossless mode it is simply encoded at maximum quality.)
     */
    private fun resolveFormat(requested: String): String =
        if (requested == "heic") "jpeg" else requested

    // ---- decode ------------------------------------------------------------

    /**
     * Memory-safe decode: `inSampleSize` for a coarse downscale, then bake in the
     * EXIF orientation, then scale to the exact target. Orientation is applied
     * unconditionally so output is upright even when `keepExif` is off.
     */
    private fun decodeUpright(source: Src, src: Probe, maxW: Int?, maxH: Int?): Bitmap {
        val (tw, th) = targetDimensions(src.width, src.height, maxW, maxH)
        var sample = 1
        while (src.width / (sample * 2) >= tw && src.height / (sample * 2) >= th) sample *= 2
        var bmp = source.decode(
            BitmapFactory.Options().apply { inSampleSize = sample },
        ) ?: error("Could not decode image")
        bmp = bmp.oriented(src.orientation)
        return if (bmp.width != tw || bmp.height != th) bmp.scaledTo(tw, th) else bmp
    }

    private fun targetDimensions(w: Int, h: Int, maxW: Int?, maxH: Int?): Pair<Int, Int> {
        var fw = w.toDouble()
        var fh = h.toDouble()
        if (maxW != null && fw > maxW) { val s = maxW / fw; fw *= s; fh *= s }
        if (maxH != null && fh > maxH) { val s = maxH / fh; fw *= s; fh *= s }
        return fw.toInt().coerceAtLeast(1) to fh.toInt().coerceAtLeast(1)
    }

    /** Scale to [w]x[h], recycling the receiver. */
    private fun Bitmap.scaledTo(w: Int, h: Int): Bitmap {
        val out = Bitmap.createScaledBitmap(this, w, h, true)
        if (out !== this) recycle()
        return out
    }

    /** Bake an EXIF [orientation] into the pixels, recycling the receiver. */
    private fun Bitmap.oriented(orientation: Int): Bitmap {
        val m = Matrix()
        when (orientation) {
            ExifInterface.ORIENTATION_ROTATE_90 -> m.postRotate(90f)
            ExifInterface.ORIENTATION_ROTATE_180 -> m.postRotate(180f)
            ExifInterface.ORIENTATION_ROTATE_270 -> m.postRotate(270f)
            ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> m.postScale(-1f, 1f)
            ExifInterface.ORIENTATION_FLIP_VERTICAL -> m.postScale(1f, -1f)
            ExifInterface.ORIENTATION_TRANSPOSE -> { m.postRotate(90f); m.postScale(-1f, 1f) }
            ExifInterface.ORIENTATION_TRANSVERSE -> { m.postRotate(270f); m.postScale(-1f, 1f) }
            else -> return this
        }
        val out = Bitmap.createBitmap(this, 0, 0, width, height, m, true)
        if (out !== this) recycle()
        return out
    }

    // ---- encode ------------------------------------------------------------

    private class Encoded(val bytes: ByteArray, val quality: Int)

    /**
     * The highest-quality encode that fits [target].
     *
     * Tries the ceiling first: an image that already fits costs a **single**
     * encode instead of a full binary search. Otherwise binary-searches
     * `[minQuality, MAX_QUALITY)`. If even [minQuality] overflows, the oversized
     * result comes back so the caller can downscale and retry.
     */
    private fun fitToTarget(
        bmp: Bitmap,
        format: String,
        lossy: Boolean,
        target: Int,
        minQuality: Int,
    ): Encoded {
        if (!lossy) return Encoded(encode(bmp, format, MAX_QUALITY, lossless = false), MAX_QUALITY)

        val ceiling = encode(bmp, format, MAX_QUALITY, lossless = false)
        if (ceiling.size <= target) return Encoded(ceiling, MAX_QUALITY)

        var lo = minQuality
        var hi = MAX_QUALITY - 1
        var best: ByteArray? = null
        var bestQuality = minQuality
        while (lo <= hi) {
            val q = (lo + hi) / 2
            val bytes = encode(bmp, format, q, lossless = false)
            if (bytes.size <= target) {
                best = bytes
                bestQuality = q
                lo = q + 1
            } else {
                hi = q - 1
            }
        }
        return Encoded(
            best ?: encode(bmp, format, minQuality, lossless = false),
            bestQuality,
        )
    }

    private fun encode(bmp: Bitmap, format: String, quality: Int, lossless: Boolean): ByteArray {
        val bos = ByteArrayOutputStream()
        bmp.compress(compressFormat(format, lossless), quality, bos)
        return bos.toByteArray()
    }

    @Suppress("DEPRECATION")
    private fun compressFormat(format: String, lossless: Boolean): Bitmap.CompressFormat = when (format) {
        "png" -> Bitmap.CompressFormat.PNG
        "webp" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            if (lossless) Bitmap.CompressFormat.WEBP_LOSSLESS else Bitmap.CompressFormat.WEBP_LOSSY
        } else {
            Bitmap.CompressFormat.WEBP
        }
        else -> Bitmap.CompressFormat.JPEG
    }

    private fun ext(format: String) = when (format) {
        "png" -> "png"; "webp" -> "webp"; else -> "jpg"
    }

    /**
     * Copy the common EXIF tags into the JPEG output. Orientation is reset to
     * normal because the rotation is already baked into the pixels — copying the
     * source tag would rotate the image a second time.
     */
    private fun copyExif(src: String, dst: String) {
        runCatching {
            val from = ExifInterface(src)
            val to = ExifInterface(dst)
            for (tag in EXIF_TAGS) from.getAttribute(tag)?.let { to.setAttribute(tag, it) }
            to.setAttribute(
                ExifInterface.TAG_ORIENTATION,
                ExifInterface.ORIENTATION_NORMAL.toString(),
            )
            to.saveAttributes()
        }.onFailure {
            // Best-effort by design — a photo without its EXIF is still a valid
            // result, so this must not fail the compression. But swallowing it
            // silently left `keepExif: true` looking like it worked, with no way
            // to tell why the metadata vanished.
            Log.w(TAG, "keepExif: could not copy EXIF from $src", it)
        }
    }

    private companion object {
        const val TAG = "FlutterCompress"
        const val MIN_QUALITY = 1
        const val MAX_QUALITY = 100
        const val MAX_DOWNSCALES = 5
        const val MIN_SIDE = 32

        /**
         * Tags copied when `keepExif` is set.
         *
         * `android.media.ExifInterface` has no "copy everything" call, so this
         * list *is* the feature's scope — anything missing here is silently lost,
         * which previously meant every exposure/lens field. Constants are inlined
         * String literals, so naming a tag the running OS predates is harmless:
         * `getAttribute` just returns null and it is skipped.
         */
        val EXIF_TAGS = listOf(
            // Camera & lens
            ExifInterface.TAG_MAKE,
            ExifInterface.TAG_MODEL,
            ExifInterface.TAG_SOFTWARE,
            ExifInterface.TAG_F_NUMBER,
            ExifInterface.TAG_APERTURE_VALUE,
            ExifInterface.TAG_MAX_APERTURE_VALUE,
            ExifInterface.TAG_EXPOSURE_TIME,
            ExifInterface.TAG_EXPOSURE_BIAS_VALUE,
            ExifInterface.TAG_EXPOSURE_PROGRAM,
            ExifInterface.TAG_EXPOSURE_MODE,
            ExifInterface.TAG_SHUTTER_SPEED_VALUE,
            ExifInterface.TAG_BRIGHTNESS_VALUE,
            ExifInterface.TAG_FOCAL_LENGTH,
            ExifInterface.TAG_FOCAL_LENGTH_IN_35MM_FILM,
            ExifInterface.TAG_FLASH,
            ExifInterface.TAG_METERING_MODE,
            ExifInterface.TAG_WHITE_BALANCE,
            @Suppress("DEPRECATION") ExifInterface.TAG_ISO_SPEED_RATINGS,
            ExifInterface.TAG_SENSING_METHOD,
            ExifInterface.TAG_DIGITAL_ZOOM_RATIO,
            ExifInterface.TAG_SCENE_CAPTURE_TYPE,
            ExifInterface.TAG_CONTRAST,
            ExifInterface.TAG_SATURATION,
            ExifInterface.TAG_SHARPNESS,
            // Timestamps
            ExifInterface.TAG_DATETIME,
            ExifInterface.TAG_DATETIME_ORIGINAL,
            ExifInterface.TAG_DATETIME_DIGITIZED,
            ExifInterface.TAG_SUBSEC_TIME,
            ExifInterface.TAG_SUBSEC_TIME_ORIGINAL,
            ExifInterface.TAG_SUBSEC_TIME_DIGITIZED,
            // Location
            ExifInterface.TAG_GPS_LATITUDE,
            ExifInterface.TAG_GPS_LATITUDE_REF,
            ExifInterface.TAG_GPS_LONGITUDE,
            ExifInterface.TAG_GPS_LONGITUDE_REF,
            ExifInterface.TAG_GPS_ALTITUDE,
            ExifInterface.TAG_GPS_ALTITUDE_REF,
            ExifInterface.TAG_GPS_TIMESTAMP,
            ExifInterface.TAG_GPS_DATESTAMP,
            ExifInterface.TAG_GPS_PROCESSING_METHOD,
            ExifInterface.TAG_GPS_SPEED,
            ExifInterface.TAG_GPS_SPEED_REF,
            ExifInterface.TAG_GPS_IMG_DIRECTION,
            ExifInterface.TAG_GPS_IMG_DIRECTION_REF,
            // Authorship / description
            ExifInterface.TAG_IMAGE_DESCRIPTION,
            ExifInterface.TAG_ARTIST,
            ExifInterface.TAG_COPYRIGHT,
            ExifInterface.TAG_USER_COMMENT,
        )
    }
}

File: darwin/flutter_compress_pro/Sources/flutter_compress_pro/ImageEngine.swift
import CoreGraphics
import Foundation
import ImageIO

/// Native image compression via ImageIO (CGImageSource/Destination). No FFmpeg,
/// no Dart software decode. For a target size it searches the quality and
/// downscales if needed — fast, so it lands accurately at/under the target.
enum ImageEngine {

  private static let minQuality = 1
  private static let maxQuality = 100
  private static let maxDownscales = 5
  private static let minSide = 32

  static func info(path: String) throws -> [String: Any] {
    guard let src = CGImageSourceCreateWithURL(URL(fileURLWithPath: path) as CFURL, nil) else {
      throw err("Cannot read image")
    }
    let meta = try probe(src)
    let size = (try? FileManager.default.attributesOfItem(atPath: path)[.size] as? Int64) ?? 0
    return [
      "path": path,
      "width": meta.width,
      "height": meta.height,
      "sizeBytes": size,
      "format": meta.format,
    ]
  }

  /// What one encode produced, before it is either written to disk or handed
  /// back as bytes.
  private struct EncodeOutcome {
    let data: Data
    let width: Int
    let height: Int
    let format: String
    let keptOriginal: Bool
    let sourceWidth: Int
    let sourceHeight: Int
    let sourceFormat: String
  }

  static func compress(
    path: String, config: ImageConfig, outputDir: String?, outputName: String?
  ) throws -> [String: Any] {
    guard let src = CGImageSourceCreateWithURL(URL(fileURLWithPath: path) as CFURL, nil) else {
      throw err("Cannot read image")
    }
    let originalSize =
      (try? FileManager.default.attributesOfItem(atPath: path)[.size] as? Int64) ?? 0
    let out = try encode(src, originalSize, config)
    if out.keptOriginal {
      return [
        "outputPath": path,
        "originalSizeBytes": originalSize,
        "compressedSizeBytes": originalSize,
        "width": out.sourceWidth,
        "height": out.sourceHeight,
        "format": out.sourceFormat,
        "skipped": true,
      ]
    }
    let dest = PluginFiles.resolveOutput(
      outputDir: outputDir, outputName: outputName, sourcePath: path, ext: ext(out.format))
    try? FileManager.default.createDirectory(
      at: dest.deletingLastPathComponent(), withIntermediateDirectories: true)
    try out.data.write(to: dest)
    return [
      "outputPath": dest.path,
      "originalSizeBytes": originalSize,
      "compressedSizeBytes": Int64(out.data.count),
      "width": out.width,
      "height": out.height,
      "format": out.format,
      "skipped": false,
    ]
  }

  /// In-memory variant: no file is read or written. `skipped` hands the caller's
  /// own bytes straight back.
  static func compressBytes(data source: Data, config: ImageConfig) throws -> [String: Any] {
    guard let src = CGImageSourceCreateWithData(source as CFData, nil) else {
      throw err("Cannot read image")
    }
    let out = try encode(src, Int64(source.count), config)
    return [
      "bytes": out.keptOriginal ? source : out.data,
      "originalSizeBytes": Int64(source.count),
      "width": out.keptOriginal ? out.sourceWidth : out.width,
      "height": out.keptOriginal ? out.sourceHeight : out.height,
      "format": out.keptOriginal ? out.sourceFormat : out.format,
      "skipped": out.keptOriginal,
    ]
  }

  private static func encode(
    _ src: CGImageSource, _ originalSize: Int64, _ config: ImageConfig
  ) throws -> EncodeOutcome {
    let source = try probe(src)

    // A nil format keeps the source's format. ImageIO can't encode WebP → JPEG;
    // JPEG stays JPEG, PNG stays PNG. (JPEG/HEIC have no lossless mode, so
    // lossless just encodes them at max quality below.)
    let requested = config.format ?? source.format
    var effective = requested == "webp" ? "jpeg" : requested
    var uti = self.uti(for: effective)
    // Comma-separated conditions short-circuit exactly like `&&`, so the probe
    // destination is only ever created for HEIC — and the brace stays on the
    // last condition's line, which `&&` across two lines does not allow.
    if effective == "heic",
      CGImageDestinationCreateWithData(NSMutableData(), uti as CFString, 1, nil) == nil {
      effective = "jpeg"
      uti = self.uti(for: "jpeg")
    }
    let lossy = effective != "png"
    // Lossless: PNG is truly lossless; JPEG/HEIC keep their format at max quality.
    let quality = config.lossless ? maxQuality : config.quality
    let exif = config.keepExif ? metadata(from: src) : nil
    // A target size can't be honored losslessly — ignore it in that case.
    let target = config.lossless ? nil : config.targetSizeKB.map { $0 * 1024 }

    // Per-axis caps, matching the Android engine: `maxWidth` bounds the width,
    // not the longest side. Aspect ratio is preserved, so asking the decoder for
    // a longest side of max(tw, th) yields exactly tw x th.
    let (tw, th) = targetDimensions(source.width, source.height, config.maxWidth, config.maxHeight)
    var image = try thumbnail(src, max(tw, th))
    var fitted = try fit(image, uti, lossy, quality, target, exif, minQuality)
    var tries = 0
    while let limit = target, fitted.data.count > limit,
      tries < maxDownscales, image.width > minSide {
      // Resample the image we already decoded instead of re-decoding the file.
      image = try resized(image, image.width * 3 / 4, image.height * 3 / 4)
      // A smaller image fits at least the quality the last round reached.
      fitted = try fit(image, uti, lossy, quality, target, exif, fitted.quality)
      tries += 1
    }
    let data = fitted.data

    // Re-encoding can end up larger than the source (already-compressed input,
    // or lossless). If so, hand back the untouched original.
    let keptOriginal = SizeMath.keepsOriginal(
      compressedBytes: Int64(data.count), originalBytes: originalSize,
      keepOriginalIfLarger: config.keepOriginalIfLarger,
      minSavingsPercent: config.minSavingsPercent)

    return EncodeOutcome(
      data: data, width: image.width, height: image.height, format: effective,
      keptOriginal: keptOriginal,
      sourceWidth: source.width, sourceHeight: source.height, sourceFormat: source.format)
  }

  // MARK: - Source inspection

  /// Source dimensions (already oriented) and format.
  private struct Probe {
    let width: Int
    let height: Int
    let format: String
  }

  private static func probe(_ src: CGImageSource) throws -> Probe {
    guard let props = CGImageSourceCopyPropertiesAtIndex(src, 0, nil) as? [CFString: Any] else {
      throw err("Cannot read image properties")
    }
    let w = (props[kCGImagePropertyPixelWidth] as? Int) ?? 0
    let h = (props[kCGImagePropertyPixelHeight] as? Int) ?? 0
    guard w > 0, h > 0 else { throw err("Cannot read image dimensions") }
    // EXIF orientations 5–8 swap the axes; report what the user actually sees.
    let orientation = (props[kCGImagePropertyOrientation] as? Int) ?? 1
    let swap = orientation >= 5 && orientation <= 8
    return Probe(
      width: swap ? h : w, height: swap ? w : h, format: sourceFormat(src))
  }

  /// Detect the source image's format (used when no format is requested).
  private static func sourceFormat(_ src: CGImageSource) -> String {
    switch (CGImageSourceGetType(src) as String?)?.lowercased() {
    case let kind? where kind.contains("png"): return "png"
    case let kind? where kind.contains("webp"): return "webp"
    case let kind? where kind.contains("heic") || kind.contains("heif"): return "heic"
    default: return "jpeg"
    }
  }

  // MARK: - Decode & resize

  private static func thumbnail(_ src: CGImageSource, _ maxPixel: Int) throws -> CGImage {
    let opts: [CFString: Any] = [
      kCGImageSourceCreateThumbnailFromImageAlways: true,
      kCGImageSourceCreateThumbnailWithTransform: true,  // bake in orientation
      kCGImageSourceThumbnailMaxPixelSize: maxPixel,
    ]
    guard let img = CGImageSourceCreateThumbnailAtIndex(src, 0, opts as CFDictionary) else {
      throw err("Could not decode image")
    }
    return img
  }

  /// Resample an already-decoded image — far cheaper than decoding the file again.
  private static func resized(_ image: CGImage, _ w: Int, _ h: Int) throws -> CGImage {
    let space =
      (image.colorSpace?.model == .rgb ? image.colorSpace : nil) ?? CGColorSpaceCreateDeviceRGB()
    guard
      let ctx = CGContext(
        data: nil, width: w, height: h, bitsPerComponent: 8, bytesPerRow: 0, space: space,
        bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue)
    else { throw err("Could not create resize context") }
    ctx.interpolationQuality = .high
    ctx.draw(image, in: CGRect(x: 0, y: 0, width: w, height: h))
    guard let out = ctx.makeImage() else { throw err("Could not resize image") }
    return out
  }

  private static func targetDimensions(
    _ w: Int, _ h: Int, _ maxW: Int?, _ maxH: Int?
  ) -> (Int, Int) {
    var fw = Double(w)
    var fh = Double(h)
    if let maxW = maxW, fw > Double(maxW) {
      let scale = Double(maxW) / fw
      fw *= scale
      fh *= scale
    }
    if let maxH = maxH, fh > Double(maxH) {
      let scale = Double(maxH) / fh
      fw *= scale
      fh *= scale
    }
    return (max(1, Int(fw)), max(1, Int(fh)))
  }

  // MARK: - Encode

  private struct Encoded {
    let data: Data
    let quality: Int
  }

  // Seven positional parameters is over SwiftLint's threshold. They are the
  // complete input to one quality search and are only ever passed together from
  // two call sites; bundling them into a struct would add a type that exists
  // solely to satisfy a counter. Both comment blocks must sit *above* the doc
  // comment — anything between `///` and the declaration orphans the doc.
  // swiftlint:disable function_parameter_count
  /// The highest-quality encode that fits `target`.
  ///
  /// Tries the ceiling first: an image that already fits costs a **single**
  /// encode instead of a full binary search. Otherwise binary-searches
  /// `[minQuality, maxQuality)`. If even `minQuality` overflows, the oversized
  /// result comes back so the caller can downscale and retry.
  private static func fit(
    _ image: CGImage, _ uti: String, _ lossy: Bool, _ quality: Int, _ target: Int?,
    _ exif: [CFString: Any]?, _ minQuality: Int
  ) throws -> Encoded {
    guard lossy, let target = target else {
      let q = lossy ? quality : maxQuality
      return Encoded(data: try encode(image, uti, q, exif), quality: q)
    }
    let ceiling = try encode(image, uti, maxQuality, exif)
    if ceiling.count <= target { return Encoded(data: ceiling, quality: maxQuality) }

    var lo = minQuality
    var hi = maxQuality - 1
    var best: Data?
    var bestQuality = minQuality
    while lo <= hi {
      let q = (lo + hi) / 2
      let encoded = try encode(image, uti, q, exif)
      if encoded.count <= target {
        best = encoded
        bestQuality = q
        lo = q + 1
      } else {
        hi = q - 1
      }
    }
    if let best = best { return Encoded(data: best, quality: bestQuality) }
    return Encoded(data: try encode(image, uti, minQuality, exif), quality: minQuality)
  }
  // swiftlint:enable function_parameter_count

  private static func encode(
    _ image: CGImage, _ uti: String, _ quality: Int, _ exif: [CFString: Any]?
  ) throws -> Data {
    let out = NSMutableData()
    guard let dest = CGImageDestinationCreateWithData(out, uti as CFString, 1, nil) else {
      throw err("No encoder available for \(uti)")
    }
    var props: [CFString: Any] = [
      kCGImageDestinationLossyCompressionQuality: Double(quality) / 100.0
    ]
    if let exif = exif { props.merge(exif) { existing, _ in existing } }
    CGImageDestinationAddImage(dest, image, props as CFDictionary)
    // Never return empty data: that would be written out as a 0-byte file and
    // reported as a successful 100% compression.
    guard CGImageDestinationFinalize(dest), out.length > 0 else {
      throw err("Image encoding failed")
    }
    return out as Data
  }

  private static func metadata(from src: CGImageSource) -> [CFString: Any] {
    let all = CGImageSourceCopyPropertiesAtIndex(src, 0, nil) as? [CFString: Any] ?? [:]
    var out: [CFString: Any] = [:]
    if var exif = all[kCGImagePropertyExifDictionary] as? [CFString: Any] {
      // These describe the *source* pixel size; after a resize they'd be stale.
      exif.removeValue(forKey: kCGImagePropertyExifPixelXDimension)
      exif.removeValue(forKey: kCGImagePropertyExifPixelYDimension)
      out[kCGImagePropertyExifDictionary] = exif
    }
    if let gps = all[kCGImagePropertyGPSDictionary] { out[kCGImagePropertyGPSDictionary] = gps }
    // Make/Model/DateTime live here — keep them so the tag set matches Android's.
    if let tiff = all[kCGImagePropertyTIFFDictionary] { out[kCGImagePropertyTIFFDictionary] = tiff }
    return out
  }

  private static func uti(for format: String) -> String {
    switch format {
    case "png": return "public.png"
    case "heic": return "public.heic"
    default: return "public.jpeg"
    }
  }

  private static func ext(_ format: String) -> String {
    switch format {
    case "png": return "png"
    case "heic": return "heic"
    default: return "jpg"
    }
  }

  private static func err(_ msg: String) -> NSError {
    NSError(domain: "flutter_compress_pro", code: 3, userInfo: [NSLocalizedDescriptionKey: msg])
  }
}
